Shell/Shell.cpp

```diff
@@ -572,10 +567,27 @@ RefPtr<Job> Shell::run_command(const AST::Command& command)
         ASSERT_NOT_REACHED();
     }
 
+    bool is_first = !command.pipeline || (command.pipeline && command.pipeline->pgid == -1);
+
+    if (command.pipeline) {
+        if (is_first) {
+            command.pipeline->pgid = child;
+        }
+    }
+
+    pid_t pgid = is_first ? child : (command.pipeline ? command.pipeline->pgid : child);
+    if (setpgid(child, pgid) < 0)
+        perror("setpgid");
+
+    if (command.should_wait) {
+        tcsetpgrp(STDOUT_FILENO, pgid);
```

Comment:
You unconditionally assign a new process group to the new children. This is fine.

You conditionally allow those children to write to the terminal, only if they run in the background? That doesn't sound right to me. Did I misread the logic here?

Reply:
The shell immediately gets control (and the terminal) back if `!command.should_wait`.

Reply:
(that is equivalent to `command&`)

Reply:
Oh! Then I just got my logic mixed up. (I thought `command.should_wait` means that the *command* should "wait" by going to background.)

Reply:
ouch.
funny story, it was originally called `should_be_run_in_background` or something like that.